compiler/src/dmd/expression.d

```diff
@@ -7101,7 +7101,7 @@ extern (C++) final class FileInitExp : DefaultInitExp
 
         Expression e = new StringExp(loc, s.toDString());
         e = e.expressionSemantic(sc);
-        e = e.castTo(sc, type);
+        e.type = Type.tstring;
```

Comment:
Type should probably be set in the constructor rather than here.

Reply:
Calling expression semantic and then manually setting the type is pointless. I think that simply calling expression semantic should fix the issue. The cast seems useless.

I see that you simply followed the existing logic, but that seems broken to me.

Reply:
Note that this code works without this patch:

```d
void foo(const(char)* mod)() {
}

void main() {
    enum string t = __MODULE__;
    foo!t();
}
```